### PR TITLE
Decouple implementation of Array and LogicArray

### PR DESCRIPTION
**Pull request by ktbarrett**
Decouples the implementation of `Array` and `LogicArray`. We create a Protocol for ArrayLike types and move a lot of common functionality into that class. We duplicate the implementation of Array into LogicArray with the anticipation that the implementation of LogicArray will change in the near future for performance reasons.

### CODE REVIEW

File: src/cocotb/types/__init__.py [hunk 1]
```diff
@@ -46,3 +64,117 @@ def concat(a: typing.Any, b: typing.Any) -> typing.Any:
...
+
+    @property
+    @abstractmethod
+    def range(self) -> Range:
+        """:class:`Range` of the indexes of the array."""
+
+    @range.setter
+    @abstractmethod
+    def range(self, new_range: Range) -> None:
+        """Sets a new indexing scheme on the array, must be the same size"""
```
Comment (cmarqu) [inline src/cocotb/types/__init__.py:96]:
```suggestion
        """Set a new indexing scheme on the array. Must be the same size."""
```

File: src/cocotb/types/__init__.py [hunk 2]
```diff
@@ -46,3 +64,117 @@ def concat(a: typing.Any, b: typing.Any) -> typing.Any:
...
+
+
+class ArrayLike(Concatable[T], Protocol, Generic[T]):
+    @property
+    def left(self) -> int:
+        """Leftmost index of the array."""
+        return self.range.left
+
+    @property
+    def direction(self) -> str:
```
Comment (cmarqu) [inline src/cocotb/types/__init__.py:79]: Maybe this could return something like a Py3.11 `StrEnum`?

File: src/cocotb/types/logic_array.py [hunk 3]
```diff
@@ -179,6 +180,40 @@ def __init__(
...
             )
 
+    @property
+    def range(self) -> Range:
+        """:class:`Range` of the indexes of the array."""
+        return self._range
+
+    @range.setter
+    def range(self, new_range: Range) -> None:
+        """Sets a new indexing scheme on the array, must be the same size"""
```
Comment (cmarqu) [inline src/cocotb/types/logic_array.py:190]:
```suggestion
        """Set a new indexing scheme on the array. Must be the same size."""
```

File: src/cocotb/types/__init__.py (hunk 2 above)
Comment (ktbarrett) [reply src/cocotb/types/__init__.py:79]: Something like this?

```python
class Direction(str):
    def __new__(cls: Type[Self], value: str) -> Self:
        value = value.lower()
        if value not in ("to", "downto"):
            raise ValueError(f"{value!r} not one of the possible enumerations ('to', 'downto')")
        return super().__new__(cls, value)

TO = Direction("to")
DOWNTO = Direction("downto")
```

String interning means we don't have to work about singletons and whatnot.

Comment (cmarqu) [reply src/cocotb/types/__init__.py:79]: Not sure if it's worth the effort, but maybe it helps catching some problems earlier.
About identity... really no idea :)

Comment (ktbarrett) [reply src/cocotb/types/__init__.py:79]: It might be, but that's definitely a follow-up PR.

**Review by marlonjames — APPROVED** [review]

**Pull request merged**